libraries/core_libs/network/include/network/tarcap/packets_handler.hpp

```diff
@@ -20,19 +20,43 @@ class PacketsHandler {
    * @param packet_type
    * @return reference to std::shared_ptr<PacketsHandler>
    */
-  std::shared_ptr<PacketHandler>& getSpecificHandler(SubprotocolPacketType packet_type);
+  const std::shared_ptr<PacketHandler>& getSpecificHandler(SubprotocolPacketType packet_type) const;
 
   /**
-   * @brief Registers handler for specific packet type
+   * @brief templated getSpecificHandler method for getting specific packet handler based on
+   * PacketHandlerType::k_packet_type
+   * @tparam PacketHandlerType
    *
-   * @param packet_type
-   * @param handler
+   * @return std::shared_ptr<PacketHandlerType>
    */
-  void registerHandler(SubprotocolPacketType packet_type, std::shared_ptr<PacketHandler> handler);
+  template <typename PacketHandlerType>
+  std::shared_ptr<PacketHandlerType> getSpecificHandler();
+
+  /**
+   * @brief Registers packet handler
+   *
+   * @tparam PacketHandlerType
+   * @tparam Args
+   * @param args to be forwarded to the PacketHandlerType ctor
+   */
+  template <typename PacketHandlerType, typename... Args>
+  void registerHandler(Args&&... args);
 
  private:
   // Map of all packets handlers, factory method selects specific packet handler for processing based on packet type
   std::unordered_map<SubprotocolPacketType, std::shared_ptr<PacketHandler>> packets_handlers_;
 };
 
+template <typename PacketHandlerType>
+std::shared_ptr<PacketHandlerType> PacketsHandler::getSpecificHandler() {
+  return std::static_pointer_cast<PacketHandlerType>(getSpecificHandler(PacketHandlerType::k_packet_type));
```

Comment:
hmm so are we agreed on k_something_something ?

Reply:
Aah I forgot, (yeah I still think its bullshit to have different styles for variables :D) but you guys decided on slack poll so democracy it is - will refactor style...